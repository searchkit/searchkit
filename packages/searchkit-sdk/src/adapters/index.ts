import { SearchkitConfig } from '..'

export interface SearchkitAdapterConstructor {
  new (config: SearchkitConfig): SearchkitAdapter
}

export interface SearchkitAdapter {
  performRequest(requestBody): Promise<any>
}
