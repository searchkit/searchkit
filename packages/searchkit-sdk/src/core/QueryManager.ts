import { SortingOption } from '..'

export type ValueFilter = {
  identifier: string
  value: string
}

export type RangeFilter = {
  identifier: string
  min?: number
  max?: number
}

export type DateRangeFilter = {
  identifier: string
  dateMin?: string
  dateMax?: string
}

export type GeoPoint = {
  lat: number
  lon: number
}

export type GeoBoundingBoxFilter = {
  identifier: string
  geoBoundingBox: {
    topLeft?: GeoPoint
    bottomRight?: GeoPoint
    topRight?: GeoPoint
    bottomLeft?: GeoPoint
  }
}

export type HierarchicalValueFilter = {
  identifier: string
  value: string
  level: number
}

export type QueryOptions = {
  fields: Array<string>
}

export type MixedFilter =
  | ValueFilter
  | RangeFilter
  | DateRangeFilter
  | GeoBoundingBoxFilter
  | HierarchicalValueFilter

export default class QueryManager {
  private sortBy: SortingOption
  private filters: Array<MixedFilter>
  private query: string
  private queryOptions: QueryOptions

  constructor() {
    this.filters = []
    this.query = null
  }

  setQuery(query: string): void {
    this.query = query
  }

  setQueryOptions(options: QueryOptions): void {
    this.queryOptions = options
  }

  setFilters(filters: Array<MixedFilter>): void {
    this.filters = filters
  }

  hasFilters(): boolean {
    return this.filters && this.filters.length > 0
  }

  hasQuery(): boolean {
    return !!(this.query && this.query.length > 0)
  }

  getQuery(): string {
    return this.query
  }

  getQueryOptions(): QueryOptions {
    return this.queryOptions
  }

  getFilters(): Array<MixedFilter> {
    return this.hasFilters() ? this.filters : []
  }

  getFiltersById(id: string): Array<MixedFilter> {
    if (!this.hasFilters()) return null
    const idFilters = this.filters.filter((filter) => filter.identifier === id)
    return idFilters.length > 0 ? idFilters : null
  }

  setSortBy(sort: SortingOption): void {
    this.sortBy = sort
  }

  getSortBy(): SortingOption {
    return this.sortBy
  }
}
