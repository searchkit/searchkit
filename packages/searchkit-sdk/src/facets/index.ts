export { default as RefinementSelectFacet } from './RefinementSelectFacet'
export { default as RangeFacet } from './RangeFacet'
export { default as DateRangeFacet } from './DateRangeFacet'
export { default as HierarchicalMenuFacet } from './HierarchicalMenuFacet'
export { default as MultiQueryOptionsFacet } from './MultiQueryOptionsFacet'
export { default as GeoDistanceOptionsFacet } from './GeoDistanceOptionsFacet'
export * from './BaseFacet'
export * from './VisibilityRules'
