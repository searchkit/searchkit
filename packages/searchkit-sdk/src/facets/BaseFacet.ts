import QueryManager, { MixedFilter } from '../core/QueryManager'

export interface FacetOptions {
  size: number
  query: string
}

export interface SelectedFilter {
  type: string
  id: string
  identifier: string
  label: string
  dateMin?: string
  dateMax?: string
  display: string
}

export interface FacetResponseEntry {
  label: string
  count: number
}

export interface FacetResponse {
  identifier: string
  label: string
  type: string
  display: string
  entries?: FacetResponseEntry[]
}

export interface BaseFacet {
  excludeOwnFilters: boolean
  getIdentifier(): string
  getLabel(): string
  getAggregation(overrides: FacetOptions, queryManager: QueryManager): any
  getFilters(filters: Array<MixedFilter>): any
  transformResponse(response: any, queryManager: QueryManager): FacetResponse
  getSelectedFilter(filterSet: any): SelectedFilter
}
