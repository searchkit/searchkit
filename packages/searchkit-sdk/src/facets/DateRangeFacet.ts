import isUndefined from 'lodash/isUndefined'
import { DateRangeFilter } from '../core/QueryManager'
import { BaseFacet } from './BaseFacet'

interface DateRangeFacetConfig {
  identifier: string
  field: string
  label: string
  display?: 'DateRange' | string
}

class DateRangeFacet implements BaseFacet {
  public excludeOwnFilters = true

  constructor(public config: DateRangeFacetConfig) {}
  getLabel(): string {
    return this.config.label
  }

  getIdentifier() {
    return this.config.identifier
  }

  getFilters(filters: Array<DateRangeFilter>) {
    const rangeFilter: { gte?: string; lte?: string } = {}
    if (!isUndefined(filters[0].dateMin)) rangeFilter.gte = filters[0].dateMin
    if (!isUndefined(filters[0].dateMax)) rangeFilter.lte = filters[0].dateMax
    return { range: { [this.config.field]: rangeFilter } }
  }

  getAggregation() {
    return {}
  }

  getSelectedFilter(filterSet) {
    return {
      type: 'DateRangeSelectedFilter',
      id: `${this.getIdentifier()}_${filterSet.dateMin}_${filterSet.dateMax}`,
      identifier: this.getIdentifier(),
      label: this.getLabel(),
      dateMin: filterSet.dateMin,
      dateMax: filterSet.dateMax,
      display: this.config.display || 'DateRangeFacet'
    }
  }

  transformResponse() {
    return {
      identifier: this.getIdentifier(),
      label: this.getLabel(),
      type: 'DateRangeFacet',
      display: this.config.display || 'DateRangeFacet',
      entries: null
    }
  }
}

export default DateRangeFacet
