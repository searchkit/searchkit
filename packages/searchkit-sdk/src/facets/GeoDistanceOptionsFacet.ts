import { ValueFilter } from '../core/QueryManager'
import { BaseFacet } from './BaseFacet'

interface GeoDistanceOptionsFacetConfig {
  identifier: string
  field: string
  origin: string
  unit?: string
  label: string
  display?: string
  ranges: {
    from?: number
    to?: number
    label: string
  }[]
  multipleSelect?: boolean
}

class GeoDistanceOptionsFacet implements BaseFacet {
  public excludeOwnFilters = false
  private unit: string

  constructor(public config: GeoDistanceOptionsFacetConfig) {
    this.excludeOwnFilters = config.multipleSelect
    this.unit = config.unit || 'mi'
  }
  getLabel(): string {
    return this.config.label
  }

  getIdentifier(): string {
    return this.config.identifier
  }

  getFilters(filterValues: ValueFilter[]) {
    const condition = this.excludeOwnFilters ? 'should' : 'must'
    const filters = filterValues.map(({ value }) => {
      const range = this.config.ranges.find((option) => option.label === value)
      return range
    })
    return {
      bool: {
        [condition]: filters.map((range) => ({
          bool: {
            ...(range.from
              ? {
                  must_not: [
                    {
                      geo_distance: {
                        distance: range.from + this.config.unit,
                        [this.config.field]: this.config.origin
                      }
                    }
                  ]
                }
              : {}),
            ...(range.to
              ? {
                  must: [
                    {
                      geo_distance: {
                        distance: range.to + this.config.unit,
                        [this.config.field]: this.config.origin
                      }
                    }
                  ]
                }
              : {})
          }
        }))
      }
    }
  }

  getAggregation() {
    return {
      [this.getIdentifier()]: {
        geo_distance: {
          field: this.config.field,
          origin: this.config.origin,
          unit: this.unit,
          keyed: true,
          ranges: this.config.ranges.map((range) => ({
            from: range.from,
            to: range.to,
            key: range.label
          }))
        }
      }
    }
  }

  getSelectedFilter(filterSet: ValueFilter) {
    return {
      identifier: this.getIdentifier(),
      id: `${this.getIdentifier()}_${filterSet.value}`,
      label: this.getLabel(),
      display: this.config.display || 'ListFacet',
      type: 'GeoDistanceOptionsFacet',
      value: filterSet.value
    }
  }

  transformResponse(response: any) {
    return {
      identifier: this.getIdentifier(),
      label: this.getLabel(),
      display: this.config.display || 'ListFacet',
      type: 'GeoDistanceOptionsFacet',
      entries: this.config.ranges.map((option) => {
        const docCount = response.buckets[option.label]?.doc_count || 0
        return {
          label: option.label,
          count: docCount
        }
      })
    }
  }
}

export default GeoDistanceOptionsFacet
