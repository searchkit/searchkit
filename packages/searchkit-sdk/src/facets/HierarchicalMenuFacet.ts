import QueryManager, { HierarchicalValueFilter } from '../core/QueryManager'
import { BaseFacet } from './BaseFacet'

interface HierarchicalMenuFacetConfig {
  identifier: string
  fields: Array<string>
  label: string
  display?: 'HierarchicalMenu' | string
}

class HierarchicalMenuFacet implements BaseFacet {
  public excludeOwnFilters = true

  constructor(public config: HierarchicalMenuFacetConfig) {}
  getLabel(): string {
    return this.config.label
  }

  getIdentifier() {
    return this.config.identifier
  }

  getFilters(filters: Array<HierarchicalValueFilter>) {
    return {
      bool: {
        must: filters.map((filter) => ({
          term: { [this.config.fields[filter.level - 1]]: filter.value }
        }))
      }
    }
  }

  getAggregation(overrides, queryManager: QueryManager) {
    const appliedFilters =
      (queryManager.getFiltersById(this.config.identifier) as Array<HierarchicalValueFilter>) || []
    const levelAggs = this.config.fields.reduce((aggs, field, index) => {
      const level = index + 1
      const parentFilters = appliedFilters.filter((f) => f.level < level)
      const getAggs = parentFilters.length === level - 1
      if (getAggs) {
        return {
          ...aggs,
          [`lvl_${level}`]: {
            filter:
              parentFilters.length === 0
                ? { match_all: {} }
                : {
                    bool: {
                      must: parentFilters.map((f) => ({
                        term: {
                          [this.config.fields[f.level - 1]]: {
                            value: f.value
                          }
                        }
                      }))
                    }
                  },
            aggs: {
              aggs: {
                terms: {
                  field: this.config.fields[level - 1]
                }
              }
            }
          }
        }
      }
      return aggs
    }, {})

    return {
      [this.getIdentifier()]: {
        filter: {
          match_all: {}
        },
        aggs: levelAggs
      }
    }
  }

  getSelectedFilter(filterSet) {
    return {
      type: 'HierarchicalValueSelectedFilter',
      id: `${this.getIdentifier()}_${filterSet.value}`,
      identifier: this.getIdentifier(),
      label: this.getLabel(),
      value: filterSet.value,
      level: filterSet.level,
      display: this.config.display || 'HierarchicalMenuFacet'
    }
  }

  transformResponse(response, queryManager: QueryManager) {
    const appliedFilters =
      (queryManager.getFiltersById(this.config.identifier) as Array<HierarchicalValueFilter>) || []
    const buildEntries = (level: number, parentId = '') => {
      if (response[`lvl_${level}`]) {
        const levelFilter = appliedFilters.find((f) => f.level === level)
        return response[`lvl_${level}`].aggs.buckets.map((bucket) => {
          const isSelected = levelFilter?.value === bucket.key
          const id = `${parentId}_${this.getIdentifier()}_${bucket.key}_${level}${
            isSelected && '_selected'
          }`
          return {
            label: bucket.key,
            count: bucket.doc_count,
            level: level,
            entries: isSelected ? buildEntries(level + 1, id) : null
          }
        })
      }
      return null
    }

    return {
      identifier: this.getIdentifier(),
      label: this.getLabel(),
      type: 'HierarchicalMenuFacet',
      display: this.config.display || 'HierarchicalMenuFacet',
      entries: buildEntries(1)
    }
  }
}

export default HierarchicalMenuFacet
