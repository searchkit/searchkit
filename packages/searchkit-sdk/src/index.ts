import type { SearchInnerHits, SearchRequest } from '@elastic/elasticsearch-types/lib/api/types'
import QueryManager, { MixedFilter, QueryOptions } from './core/QueryManager'
import RequestBodyBuilder from './core/RequestBodyBuilder'
import BaseQuery from './query/BaseQuery'
import { BaseFacet } from './facets/BaseFacet'
import { BaseFilter } from './filters/BaseFilter'
import { VisibleWhenRuleSet } from './facets'
import { SearchkitTransporter, FetchClientTransporter } from './transporters'
import { getAggregationsFromFacets } from './core/FacetsFns'
import {
  SearchkitResponseTransformer,
  ElasticSearchResponseTransformer,
  SearchkitResponse,
  SearchkitHit
} from './transformers'
import { BaseSuggestor } from './suggestors'

export * from './query'
export * from './facets'
export * from './filters'
export * from './transporters'
export * from './suggestors'
export { MixedFilter, QueryOptions, SearchkitResponse, SearchkitHit }

export interface SortingOption {
  id: string
  label: string
  field: any
  defaultOption?: boolean
}

export interface CustomHighlightConfig {
  field: string
  config: any
}

export interface CloudHost {
  id: string
}

export interface BaseConfig {
  host?: string
  cloud?: CloudHost
  index: string
  connectionOptions?: {
    apiKey?: string
    headers?: Record<string, string>
  }
}

export type CollapseInnerHitsConfig = SearchInnerHits

export interface CollapseConfig {
  field: string
  inner_hits?: CollapseInnerHitsConfig | CollapseInnerHitsConfig[]
}

export interface SearchkitConfig extends BaseConfig {
  suggestions?: Array<BaseSuggestor<any>>
  sortOptions?: SortingOption[]
  hits?: {
    fields: string[]
    highlightedFields?: (string | CustomHighlightConfig)[]
  }
  query?: BaseQuery
  facets?: Array<BaseFacet | VisibleWhenRuleSet>
  filters?: Array<BaseFilter>
  collapse?: CollapseConfig
  postProcessRequest?: (body: SearchRequest) => SearchRequest
}

export interface ResultsResolverParameters {
  filters: Array<MixedFilter>
  query: string
}

const getFacets = (
  facets: Array<BaseFacet | VisibleWhenRuleSet> = [],
  queryManager: QueryManager,
  ctx
) =>
  facets.reduce((facetsList, facet) => {
    if (facet instanceof VisibleWhenRuleSet) {
      return [...facetsList, ...facet.getActiveFacets(queryManager, ctx)]
    }
    return [...facetsList, facet]
  }, [])

const createInstance = (
  config: SearchkitConfig,
  transporter?: SearchkitTransporter
): SearchkitRequest => new SearchkitRequest(config, transporter)

export type ResponseRequest = {
  hits?: {
    size?: number
    from?: number
    includeRawHit?: boolean
  }
  facets?:
    | boolean
    | Array<{
        identifier: string
        query?: string
        size?: number
      }>
}

export type BaseFilters = Array<any>

function getSortOption(id, sortOptions: SortingOption[]) {
  const selectedSortOption = sortOptions.find((sortOption) => sortOption.id === id)
  if (!selectedSortOption) {
    throw new Error(`Sort Option: sorting option ${id} not found`)
  }
  return selectedSortOption
}

export class SearchkitRequest {
  private queryManager: QueryManager
  private transporter: SearchkitTransporter
  private transformer: SearchkitResponseTransformer

  constructor(private config: SearchkitConfig, transporter?: SearchkitTransporter) {
    this.queryManager = new QueryManager()
    this.transporter = !transporter ? new FetchClientTransporter(config) : transporter
    this.transformer = new ElasticSearchResponseTransformer()
    this.handleDefaults()
  }

  handleDefaults(): void {
    const defaultSortOptionId =
      this.config.sortOptions && this.config.sortOptions.find((option) => option.defaultOption)
    const defaultSortOption =
      defaultSortOptionId && getSortOption(defaultSortOptionId.id, this.config.sortOptions)
    this.queryManager.setSortBy(defaultSortOption)
  }

  query(query: string): SearchkitRequest {
    this.queryManager.setQuery(query)
    return this
  }

  setFilters(filters: Array<MixedFilter>): SearchkitRequest {
    this.queryManager.setFilters(filters)
    return this
  }

  setQueryOptions(options: QueryOptions): SearchkitRequest {
    this.queryManager.setQueryOptions(options)
    return this
  }

  setSortBy(sortBy: string): SearchkitRequest {
    if (this.config.sortOptions && sortBy) {
      const sortOption = getSortOption(sortBy, this.config.sortOptions)
      this.queryManager.setSortBy(sortOption)
    }
    return this
  }

  async executeSuggestions(query: string): Promise<Array<any>> {
    const suggestions = await Promise.all(
      this.config.suggestions.map((suggestor) => suggestor.getSuggestions(query, this.transporter))
    )
    return suggestions
  }

  async execute(
    responseRequest: ResponseRequest,
    baseFilters: BaseFilters = []
  ): Promise<SearchkitResponse> {
    const partialQueries = []
    const facets = getFacets(this.config.facets, this.queryManager, {})
    let filteredFacets = null
    let overrides = {}

    if (!responseRequest.hits) responseRequest.hits = {}
    if (!responseRequest.hits.size) responseRequest.hits.size = 0
    if (!responseRequest.hits.from) responseRequest.hits.from = 0

    if (responseRequest.facets) {
      if (Array.isArray(responseRequest.facets)) {
        const filteredFacetIdentifiers = responseRequest.facets.map((facet) => facet.identifier)
        filteredFacets =
          facets &&
          facets.filter((facet) => filteredFacetIdentifiers.includes(facet.getIdentifier()))
        overrides = responseRequest.facets.reduce(
          (acc, facet) => ({
            ...acc,
            [facet.identifier]: {
              query: facet.query,
              size: facet.size
            }
          }),
          {}
        )
      }
      const aggs = getAggregationsFromFacets(this.queryManager, overrides, filteredFacets || facets)
      partialQueries.push(aggs)
    }

    const chosenSortOption = this.queryManager.getSortBy()

    partialQueries.push({
      size: responseRequest.hits.size,
      from: responseRequest.hits.from,
      sort: chosenSortOption ? chosenSortOption.field : [{ _score: 'desc' }]
    })

    let skRequestBody = RequestBodyBuilder(
      this.queryManager,
      this.config,
      baseFilters,
      facets,
      partialQueries
    )
    if (this.config.postProcessRequest) {
      skRequestBody = this.config.postProcessRequest(skRequestBody)
    }
    const response = await this.transporter.performRequest(skRequestBody)

    return this.transformer.transformResponse(
      response,
      filteredFacets || facets,
      this.queryManager,
      this.config,
      responseRequest
    )
  }
}

export default createInstance
