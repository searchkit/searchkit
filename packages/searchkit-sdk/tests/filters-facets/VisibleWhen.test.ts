import nock from 'nock'
import SearchkitRequest, { MultiMatchQuery, SearchkitConfig, VisibleWhen } from '../../src'
import { FacetSelectedRule, RefinementSelectFacet } from '../../src/facets'
import ResultsNoHitsMock from '../__mock-data__/Facets/results-no-hits.json'
import QueryManager from '../../src/core/QueryManager'

describe('Visible When Rule system', () => {
  it('success', async () => {
    const customRule = jest.fn((queryManager: QueryManager, ctx: any) => true)

    const config: SearchkitConfig = {
      host: 'http://localhost:9200',
      index: 'movies',
      hits: {
        fields: ['actors', 'writers']
      },
      query: new MultiMatchQuery({ fields: ['actors', 'writers', 'title^4', 'plot'] }),
      facets: [
        new RefinementSelectFacet({ identifier: 'type', field: 'type.raw', label: 'Type' }),
        VisibleWhen(
          [
            new RefinementSelectFacet({
              identifier: 'writers',
              field: 'writers.raw',
              label: 'Writers',
              display: 'override',
              multipleSelect: true
            }),
            new RefinementSelectFacet({
              identifier: 'actors',
              field: 'actors.raw',
              label: 'Actors'
            }),
            new RefinementSelectFacet({
              identifier: 'genres',
              field: 'genres.raw',
              label: 'Genres'
            })
          ],
          [FacetSelectedRule('type', 'Movie'), customRule]
        )
      ]
    }

    const request = SearchkitRequest(config)

    const scope = nock('http://localhost:9200')
      .post('/movies/_search')
      .reply(200, (uri, body) => {
        const facetFieldRequests = (body as any).aggs.facet_bucket_all.aggs
        expect(facetFieldRequests).toBeTruthy()
        expect(facetFieldRequests.type.terms.field).toBe('type.raw')
        expect(Object.keys(facetFieldRequests)).toHaveLength(1)
        return ResultsNoHitsMock
      })

    let response = await request.execute({
      facets: true
    })
    expect(response).toMatchSnapshot()
    expect(response.facets).toHaveLength(1)
    expect(response.facets.map((f) => f.identifier)).toEqual(['type'])

    nock('http://localhost:9200')
      .post('/movies/_search')
      .reply(200, (uri, body) => {
        const facetFieldRequests = (body as any).aggs.facet_bucket_all.aggs
        expect(facetFieldRequests).toBeTruthy()
        expect(facetFieldRequests.type.terms.field).toBe('type.raw')
        expect(Object.keys(facetFieldRequests).sort()).toEqual(['type', 'actors', 'genres'].sort())
        return ResultsNoHitsMock
      })

    request.setFilters([{ identifier: 'type', value: 'Movie' }])

    response = await request.execute({
      facets: true
    })
    expect(response).toMatchSnapshot()
    expect(response.facets).toHaveLength(4)
    expect(response.facets.map((f) => f.identifier).sort()).toEqual(
      ['type', 'writers', 'genres', 'actors'].sort()
    )

    customRule.mockReturnValueOnce(false)

    nock('http://localhost:9200')
      .post('/movies/_search')
      .reply(200, (uri, body) => ResultsNoHitsMock)

    response = await request.execute({
      facets: true
    })
    expect(response).toMatchSnapshot()
    expect(response.facets).toHaveLength(1)
    expect(response.facets.map((f) => f.identifier).sort()).toEqual(['type'].sort())
  })
})
