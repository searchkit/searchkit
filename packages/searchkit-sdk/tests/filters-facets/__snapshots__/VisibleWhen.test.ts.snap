// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Visible When Rule system success 1`] = `
Object {
  "facets": Array [
    Object {
      "display": "ListFacet",
      "entries": Array [
        Object {
          "count": 3120,
          "label": "Movie",
        },
        Object {
          "count": 521,
          "label": "Episode",
        },
        Object {
          "count": 473,
          "label": "Series",
        },
        Object {
          "count": 48,
          "label": "Game",
        },
      ],
      "identifier": "type",
      "label": "Type",
      "type": "RefinementSelectFacet",
    },
  ],
  "items": Array [],
  "page": Object {
    "from": 0,
    "pageNumber": NaN,
    "size": 0,
    "total": 4162,
    "totalPages": Infinity,
  },
  "sortedBy": undefined,
  "summary": Object {
    "appliedFilters": Array [],
    "disabledFilters": Array [],
    "query": "",
    "sortOptions": Array [],
    "total": 4162,
  },
}
`;

exports[`Visible When Rule system success 2`] = `
Object {
  "facets": Array [
    Object {
      "display": "ListFacet",
      "entries": Array [
        Object {
          "count": 3120,
          "label": "Movie",
        },
        Object {
          "count": 521,
          "label": "Episode",
        },
        Object {
          "count": 473,
          "label": "Series",
        },
        Object {
          "count": 48,
          "label": "Game",
        },
      ],
      "identifier": "type",
      "label": "Type",
      "type": "RefinementSelectFacet",
    },
    Object {
      "display": "override",
      "entries": Array [
        Object {
          "count": 83,
          "label": "J.J. Abrams",
        },
        Object {
          "count": 74,
          "label": "Jeffrey Lieber",
        },
        Object {
          "count": 73,
          "label": "Damon Lindelof",
        },
        Object {
          "count": 53,
          "label": "James Manos Jr.",
        },
        Object {
          "count": 53,
          "label": "Jeff Lindsay",
        },
      ],
      "identifier": "writers",
      "label": "Writers",
      "type": "RefinementSelectFacet",
    },
    Object {
      "display": "ListFacet",
      "entries": Array [
        Object {
          "count": 73,
          "label": "Naveen Andrews",
        },
        Object {
          "count": 56,
          "label": "Jennifer Carpenter",
        },
        Object {
          "count": 56,
          "label": "Michael C. Hall",
        },
        Object {
          "count": 53,
          "label": "Emilie de Ravin",
        },
        Object {
          "count": 42,
          "label": "Jared Padalecki",
        },
      ],
      "identifier": "actors",
      "label": "Actors",
      "type": "RefinementSelectFacet",
    },
    Object {
      "display": "ListFacet",
      "entries": Array [
        Object {
          "count": 2529,
          "label": "Drama",
        },
        Object {
          "count": 1274,
          "label": "Comedy",
        },
        Object {
          "count": 786,
          "label": "Crime",
        },
        Object {
          "count": 725,
          "label": "Action",
        },
        Object {
          "count": 666,
          "label": "Adventure",
        },
      ],
      "identifier": "genres",
      "label": "Genres",
      "type": "RefinementSelectFacet",
    },
  ],
  "items": Array [],
  "page": Object {
    "from": 0,
    "pageNumber": NaN,
    "size": 0,
    "total": 4162,
    "totalPages": Infinity,
  },
  "sortedBy": undefined,
  "summary": Object {
    "appliedFilters": Array [
      Object {
        "display": "ListFacet",
        "id": "type_Movie",
        "identifier": "type",
        "label": "Type",
        "type": "ValueSelectedFilter",
        "value": "Movie",
      },
    ],
    "disabledFilters": Array [],
    "query": "",
    "sortOptions": Array [],
    "total": 4162,
  },
}
`;

exports[`Visible When Rule system success 3`] = `
Object {
  "facets": Array [
    Object {
      "display": "ListFacet",
      "entries": Array [
        Object {
          "count": 3120,
          "label": "Movie",
        },
        Object {
          "count": 521,
          "label": "Episode",
        },
        Object {
          "count": 473,
          "label": "Series",
        },
        Object {
          "count": 48,
          "label": "Game",
        },
      ],
      "identifier": "type",
      "label": "Type",
      "type": "RefinementSelectFacet",
    },
  ],
  "items": Array [],
  "page": Object {
    "from": 0,
    "pageNumber": NaN,
    "size": 0,
    "total": 4162,
    "totalPages": Infinity,
  },
  "sortedBy": undefined,
  "summary": Object {
    "appliedFilters": Array [
      Object {
        "display": "ListFacet",
        "id": "type_Movie",
        "identifier": "type",
        "label": "Type",
        "type": "ValueSelectedFilter",
        "value": "Movie",
      },
    ],
    "disabledFilters": Array [],
    "query": "",
    "sortOptions": Array [],
    "total": 4162,
  },
}
`;
